File: code/Python/Chapter7/7-3/fast_exponentiation.py
def fast_exponentiation(x, n):
    if n == 0:
        return 1
    
    if n % 2 == 0:
        half = fast_exponentiation(x, n // 2)
        return half * half
    else:
        half = fast_exponentiation(x, (n - 1) // 2)
        return half * half * x

print(fast_exponentiation(2, 18))

File: code/Java/Chapter7/7-3/FastExponentiation.java
public class FastExponentiation {
    public static void main(String[] args) {
        System.out.println(fastExponentiation(2, 18));
    }

    public static long fastExponentiation(long x, long n) {
        if (n == 0) {
            return 1;
        }

        long half;
        if (n % 2 == 0) {
            half = fastExponentiation(x, n / 2);
            return half * half;
        } else {
            half = fastExponentiation(x, (n - 1) / 2);
            return half * half * x;
        }
    }
}